File: skills/code-architect/references/code-examples/typescript/Prototype.ts
// Prototype — copy existing objects via a clone interface without coupling to classes.

interface Cloneable<T> {
  clone(): T;
}

class Document implements Cloneable<Document> {
  constructor(
    public title: string,
    public tags: string[],
    public metadata: Record<string, string>,
  ) {}

  clone(): Document {
    // Deep-copy mutable members so clones stay independent.
    return new Document(this.title, [...this.tags], { ...this.metadata });
  }
}

function demo(): void {
  const original = new Document("Spec", ["draft"], { author: "team" });
  const copy = original.clone();
  copy.title = "Spec (revised)";
  copy.tags.push("reviewed");

  console.log(`original: ${original.title} / tags=${original.tags.join(",")}`);
  console.log(`copy:     ${copy.title} / tags=${copy.tags.join(",")}`);
}

demo();

File: skills/code-architect/references/code-examples/swift/Prototype.swift
// Prototype — copy existing objects via a clone interface without coupling to classes.

protocol Cloneable {
    func clone() -> Self
}

final class Document: Cloneable {
    var heading: String
    var tags: [String]

    init(heading: String, tags: [String]) {
        self.heading = heading
        self.tags = tags
    }

    func clone() -> Document {
        Document(heading: heading, tags: tags)
    }

    var description: String {
        "Document(heading: \"\(heading)\", tags: \(tags))"
    }
}

func runDemo() {
    let original = Document(heading: "Draft", tags: ["a", "b"])
    let copy = original.clone()
    copy.heading = "Revised Copy"
    copy.tags.append("c")

    print("original:", original.description)
    print("copy:    ", copy.description)
    print("independent:", original.heading != copy.heading)
}

runDemo()
